openstreetmap-carto.lua

```diff
@@ -0,0 +1,277 @@
+   {'highway', 'trunk', 8, 370, 1},
+   {'highway', 'primary', 7, 360, 1},
+   {'highway', 'secondary', 6, 350, 1},
+   {'highway', 'tertiary', 4, 340, 0},
+   {'highway', 'residential', 3, 330, 0},
+   {'highway', 'unclassified', 3, 330, 0},
+   {'highway', 'road', 3, 330, 0},
+   {'highway', 'living_street', 0, 320, 0},
+   {'highway', 'pedestrian', 0, 310, 0},
+   {'highway', 'raceway', 0, 300, 0},
+   {'highway', 'motorway_link', 9, 240, 1},
+   {'highway', 'trunk_link', 6, 230, 1},
+   {'highway', 'primary_link', 5, 220, 1},
+   {'highway', 'secondary_link', 4, 210, 1},
+   {'highway', 'tertiary_link', 3, 200, 0},
+   {'highway', 'service', 0, 150, 0},
+   {'highway', 'track', 0, 110, 0},
+   {'highway', 'path', 0, 100, 0},
+   {'highway', 'footway', 0, 100, 0},
+   {'highway', 'bridleway', 0, 100, 0},
+   {'highway', 'cycleway', 0, 100, 0},
+   {'highway', 'steps', 0, 100, 0},
+   {'highway', 'platform', 0, 90, 0},
+   {'highway', 'minor', 3, 0, 0},
+   {'railway', 'platform', 0, 90, 0},
+   {'aeroway', 'runway', 0, 60, 0},
+   {'aeroway', 'taxiway', 0, 50, 0},
+   {'highway', 'construction', 0, 10, 0}
+}
+
+function add_z_order(keyvalues)
+   -- The default z_order is 0
+   z_order = 0
+   osmcarto_z_order = 0
+
+   -- Increase or decrease z_order based on the specific key/value combination as specified in zordering_tags
+   for i,k in ipairs(zordering_tags) do
+      -- If the value in zordering_tags is specified, match key and value. Otherwise, match key only.
+      if ((k[2]  and keyvalues[k[1]] == k[2]) or (k[2] == nil and keyvalues[k[1]] ~= nil)) then
+         -- If the fifth component of the element of zordering_tags is 1, add the object to planet_osm_roads
+         if (k[5] == 1) then
+            roads = 1
+         end
+         z_order = z_order + k[3]
+         osmcarto_z_order = math.max(osmcarto_z_order, k[4])
+      end
+   end
+
+   -- Add z_order as key/value combination
+   keyvalues["osmcarto_z_order"] = osmcarto_z_order
+   keyvalues["z_order"] = z_order
+
+   return keyvalues, roads
+end
+
+-- Filtering on nodes, ways, and relations
+function filter_tags_generic(keyvalues, numberofkeys)
+   filter = 0   -- Will object be filtered out?
+
+   -- Filter out objects with 0 tags
+   if numberofkeys == 0 then
+      filter = 1
+      return filter, keyvalues
+   end
+
+   -- Delete tags listed in delete_tags
+   for i,k in ipairs(delete_tags) do
+      keyvalues[k] = nil
+   end
+
+   return filter, keyvalues
+end
+
+-- Filtering on nodes
+function filter_tags_node (keyvalues, numberofkeys)
+   return filter_tags_generic(keyvalues, numberofkeys)
+end
+
+-- Filtering on relations
+function filter_basic_tags_rel (keyvalues, numberofkeys)
+   -- Filter out objects that are filtered out by filter_tags_generic
+   filter, keyvalues = filter_tags_generic(keyvalues, numberofkeys)
+   if filter == 1 then
+      return filter, keyvalues
+   end
+
+   -- Filter out all relations except route, multipolygon and boundary relations
```

Comment:
Keys like “FIXME” and “note” are filtered by a blacklist. But here, for the relations, a whitelist is used. Would it be very bad for performance to put all relations in the database (maybe without some blacklisted types)? Maybe one day you want to use some of these relations, and given that database reloads seem to be seldom if ever…